src/wp-includes/blocks.php

```diff
@@ -157,24 +157,51 @@ function register_block_style_handle( $metadata, $field_name ) {
 	if ( empty( $metadata[ $field_name ] ) ) {
 		return false;
 	}
+
+	// Check whether styles should have a ".min" suffix or not.
+	$suffix = SCRIPT_DEBUG ? '' : '.min';
+
+	$is_core_block   = isset( $metadata['file'] ) && 0 === strpos( $metadata['file'], ABSPATH . WPINC );
+	$should_register = ! $is_core_block || should_load_separate_block_assets();
+
 	$style_handle = $metadata[ $field_name ];
 	$style_path   = remove_block_asset_path_prefix( $metadata[ $field_name ] );
-	if ( $style_handle === $style_path ) {
+
+	if ( $style_handle === $style_path && ! $is_core_block ) {
 		return $style_handle;
 	}
 
+	if ( ! $should_register ) {
```

Comment:
This can be moved up right after `if ( empty( $metadata[ $field_name ] ) ) {` in line 157.

From the readability standpoint inlining the check might also be better:
```php
if ( $is_core_block && ! should_load_separate_block_assets() ) {
    return false;
}
```